src/v/kafka/server/quota_manager.h

```diff
@@ -36,6 +38,53 @@
 
 namespace kafka {
 
+using client_quota_clock = ss::lowres_clock;
+
+class client_quotas_probe {
+public:
+    using clock = client_quota_clock;
+
+    client_quotas_probe() = default;
+    client_quotas_probe(const client_quotas_probe&) = delete;
+    client_quotas_probe& operator=(const client_quotas_probe&) = delete;
+    client_quotas_probe(client_quotas_probe&&) = delete;
+    client_quotas_probe& operator=(client_quotas_probe&&) = delete;
+    ~client_quotas_probe() noexcept = default;
+
+    void setup_metrics();
+
+    void record_throttle_time(
+      client_quota_rule rule, client_quota_type qt, clock::duration t);
+    void record_throughput(
+      client_quota_rule rule, client_quota_type qt, uint64_t val);
+
+private:
+    ss::metrics::histogram
+    get_throttle_time(client_quota_rule rule, client_quota_type qt) const;
+    ss::metrics::histogram
+    get_throughput(client_quota_rule rule, client_quota_type qt) const;
+
+    struct granular_metrics {
+        log_hist_client_quota throttle_time;
+        log_hist_internal throughput;
+    };
+
+    granular_metrics& operator()(client_quota_rule, client_quota_type);
+    const granular_metrics&
+    operator()(client_quota_rule, client_quota_type) const;
+
+    // Assume the enums values are in sequence: [0, all_*.size())
+    static_assert(static_cast<size_t>(all_client_quota_rules[0]) == 0);
+    static_assert(static_cast<size_t>(all_client_quota_types[0]) == 0);
+    using metrics_container_t = std::array<
+      std::array<granular_metrics, all_client_quota_types.size()>,
+      all_client_quota_rules.size()>;
+
+    metrics::internal_metric_groups _internal_metrics;
+    metrics::public_metric_groups _public_metrics;
```

Comment:
nitpick: `all_metrics_groups` (not worth changing unless something is being changed)

Reply:
I've changed this now